zebra/zebra_vty.c

```diff
@@ -1276,25 +1278,44 @@ static int show_nexthop_group_id_cmd_helper(struct vty *vty, uint32_t id)
 	return CMD_SUCCESS;
 }
 
-static void show_nexthop_group_cmd_helper(struct vty *vty,
-					  struct zebra_vrf *zvrf, afi_t afi)
+/* Helper function for iteration through the hash of nexthop-groups/nhe-s */
+
+struct nhe_show_context {
+	struct vty *vty;
+	vrf_id_t vrf_id;
+	afi_t afi;
+};
+
+static int nhe_show_walker(struct hash_bucket *bucket, void *arg)
 {
-	struct list *list = hash_to_list(zrouter.nhgs);
-	struct nhg_hash_entry *nhe = NULL;
-	struct listnode *node = NULL;
+	struct nhe_show_context *ctx = arg;
+	struct nhg_hash_entry *nhe;
 
-	for (ALL_LIST_ELEMENTS_RO(list, node, nhe)) {
+	nhe = bucket->data; /* We won't be offered NULL buckets */
 
-		if (afi && nhe->afi != afi)
-			continue;
+	if (ctx->afi && nhe->afi != ctx->afi)
+		goto done;
 
-		if (nhe->vrf_id != zvrf->vrf->vrf_id)
-			continue;
+	if (ctx->vrf_id && nhe->vrf_id != ctx->vrf_id)
+		goto done;
 
-		show_nexthop_group_out(vty, nhe);
-	}
+	show_nexthop_group_out(ctx->vty, nhe);
+
+done:
+	return HASHWALK_CONTINUE;
+}
+
+static void show_nexthop_group_cmd_helper(struct vty *vty,
+					  struct zebra_vrf *zvrf,
+					  afi_t afi)
+{
+	struct nhe_show_context ctx;
+
+	ctx.vty = vty;
+	ctx.afi = afi;
+	ctx.vrf_id = zvrf->vrf->vrf_id;
 
-	list_delete(&list);
+	hash_walk(zrouter.nhgs_id, nhe_show_walker, &ctx);
```

Comment:
I'm a bit confused. Hash tables are unordered by definition. How can `hash_walk()` iterate over the elements in ID order? I'm probably missing something here...

Reply:
doh - yes, you're quite right: I'd been spending time with something else that used a tree, and just sort of carried that over somehow.

I'd still like to avoid the complete copy of the hashtable: I'll re-word the PR...